### PR TITLE
WIP: New datasource google snapshot

### PR DESCRIPTION
**Pull request by tpoindessous**
Hi

I wanted to add a new datasource for google_compute_snapshot.

I added support for search by labels.

It's WIP for now, I need to add support for filter and add website doc.

- [X] new datasource for google_compute_snapshot
- [X] basic test
- [X] support for search by labels
- [ ] website documentation
- [ ] support for search by filter

If you have time to do a first review and give me some advice, it would be great !

Thanks !

### CODE REVIEW

File: google/data_source_google_compute_snapshot.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"source_disk_encryption_key_sha256": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"disk_size": &schema.Schema{
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: nit: can this be disk_size_gb?

File: google/data_source_google_compute_snapshot.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			"storage_size": &schema.Schema{
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: nit: can this be storage_size_bytes? (or just storage_bytes to match the api)?

File: google/data_source_google_compute_snapshot.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	d.Set("storage_size", snapshot.StorageBytes)
+	d.Set("storage_size_status", snapshot.StorageBytesStatus)
+	d.Set("disk_size", snapshot.DiskSizeGb)
+	d.Set("labels", snapshot.Labels)
+
+	d.SetId(snapshot.Name)
+	return nil
+}
+
+func resourceSnapshotLabels(d *schema.ResourceData) map[string]string {
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: Likewise with the schema one- can this be moved into a separate file? Then you can also get rid of the one defined in `resource_compute_instance.go`

File: google/data_source_google_compute_snapshot.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+			},
+
+			"self_link": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: Mind grouping all the configurable properties together at the top of the list?

File: google/data_source_google_compute_snapshot.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+		filter := ""
+		log.Printf("[DEBUG] Labels length : %d", len(labels))
+		for k, v := range labels {
+			log.Printf("[DEBUG] Label key : '%s', value : '%s'", k, v)
+			filter = fmt.Sprintf("%s(labels.%s eq %s)", filter, k, v)
+		}
+		log.Printf("[DEBUG] Labels filter : %s", filter)
+		snapshotList, err := config.clientCompute.Snapshots.List(project).Filter(filter).Do()
+		if err != nil {
+			return fmt.Errorf("Snapshot, error while Listing with filter '%s' ", filter)
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: nit: I'd rephrase this as
`return fmt.Errorf("error while listing snapshots with filter %q: %s", filter, err)`

File: google/data_source_google_compute_snapshot.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+		}
+		log.Printf("[DEBUG] SnapshotList length : %d", len(snapshotList.Items))
+
+		if len(snapshotList.Items) > 1 {
+			return fmt.Errorf("Snapshot : too many snapshots found with these tags")
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: Why is it an error to have multiple snapshots with the same set of labels?

Also `tags`->`labels`

File: google/data_source_google_compute_snapshot.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+		} else if len(snapshotList.Items) == 0 {
+			return fmt.Errorf("Snapshot : no snapshot found with these tags")
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: nit: I don't think you need the "Snapshot :" in front since snapshot is already mentioned in the error message. Also `tags` -> `labels`

File: google/data_source_google_compute_snapshot.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+		} else {
+			d.Set("name", snapshotList.Items[0].Name)
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: If both `name` and `labels` are specified, name will be overwritten. Is that intended?

File: google/data_source_google_compute_snapshot_test.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceGoogleSnapshotCheck(
+						"data.google_compute_snapshot.foobar", "google_compute_snapshot.foobar"),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceGoogleSnapshotCheck(data_source_name string, resource_name string) resource.TestCheckFunc {
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot_test.go]: prefer camelCase for variable names

File: google/data_source_google_compute_snapshot.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+		Read: dataSourceGoogleComputeSnapshotRead,
+
+		Schema: map[string]*schema.Schema{
+			//"filter": dataSourceFiltersSchema(),
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"description": &schema.Schema{
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: Should attributes not present in `resource_compute_snapshot` still be included?

File: google/data_source_google_compute_snapshot_test.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+					ds_attr[attr_to_check],
+					rs_attr[attr_to_check],
+				)
+			}
+		}
+		return nil
+	}
+}
+
+func testAccComputeDataSnapshot_basic(snapshotName string, diskName string) string {
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot_test.go]: Can you also add at least one test that tests the labels filtering?

File: google/data_source_google_compute_snapshot.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+			return fmt.Errorf("Snapshot : too many snapshots found with these tags")
+		} else if len(snapshotList.Items) == 0 {
+			return fmt.Errorf("Snapshot : no snapshot found with these tags")
+		} else {
+			d.Set("name", snapshotList.Items[0].Name)
+		}
+	}
+
+	snapshot, err := config.clientCompute.Snapshots.Get(
+		project, d.Get("name").(string)).Do()
```
Comment (danawillow) [inline google/data_source_google_compute_snapshot.go]: What if `name` is unspecified?

**Comment by tpoindessous** [comment]
Hi

thanks for your review, I will do the corrections before end of august, I hope (summer holidays !!!).

**Comment by tpoindessous** [comment]
Hi

I tried to correct my code following your remarks. I didn't finish yet.

Now, I need to finish my other PR for adding support of labels in resource snapshot, then after, I could test filtering labels.

I will inform you when this work is finished. 

Thanks !

**Comment by ndmckinley** [comment]
Hey there!  Is this still on your radar?  We're happy to leave the PR open, but I'm working on cleaning up our backlog and if it's not, I'll close it.  Just let me know!  Thanks!

**Comment by tpoindessous** [comment]
Hi,

yes, I will finish this work, I just need a little more time ...

Thanks !

Le lun. 12 mars 2018 à 21:46, Nathan McKinley <notifications@github.com> a
écrit :

**Comment by tpoindessous** [comment]
hi @ndmckinley , I don't think I will be able to finish this task soon ..

Sorry for that, but can you close this PR ?

Thanks.
